### ISSUE TITLE
Node Conformance Test: Package Node Conformance Test

### ISSUE THREAD

**Issue by Random-Liu** [kubernetes/kubernetes#30174]
For #30122.
In 1.4, we plan to provide part of the node e2e test to the user as the node conformance test.

For user adoption, We should package the node conformance test well to make it easy to ship and for the user to use. 

After discussion with @dchen1107 and @vishh, here is the current plan:
- [x] #30348 #30540 **Separate node conformance test and its dependencies from node e2e framework.** To package node conformance test, we should firstly figure out what should be included in node conformance test and separate them out. According to the UX defined in #30122, the node conformance test should include:
  - [Required] Ginkgo test suite
  - [Required] Necessary kubernetes services:
    - Apiserver
    - Etcd
    - Namespace controller #29020
  - [Nice-to-have] Ginkgo. The binary of test suite already has enough functionality to run the test sequentially. However, `ginkgo` is more powerful, e.g. we still need `ginkgo` to run test in parallel. So it is nice to have.
  - [ ] [Optional] Move kubelet start logic out of test suite. Currently in node e2e, we start kubelet in the test suite, which is not needed in the node conformance test. We should separate the logic out of the test suite into the node e2e framework.
  - [ ] [Optional] #36919 Make the current kubelet flags as test suite input and make it accessible by all tests, so that we could:
    - Filter test based on kubelet configuration.
    - Run test against current kubelet configuration.
- [x] **Make the node conformance test one binary.** To make it easier to ship and use, we decided to statically link all required services into a single test binary. To achieve this, we need to:
  - [x] #30116 Add `run-services-mode` and start dependent services in a separate process. The reason why we need to do this:
    - Keep the test output clean. Both the test and its dependencies are using `glog`, but `glog` can only be configured at process level. After we statically link all the services with the test, there's no way to distinguish their log from test result. To solve this, we need to start all services in another process and configure its log separately.
    - Keep the test a single binary. We don't want to introduce another binary to start the services, so we just add a special `run-services-mode` mode.
  - [x] Statically link services.
    - [x] #30211 Apiserver.
    - [x] #30198 Etcd.
    - [x] #30200 Namespace controller.
- [x] #31093 **Package the node conformance test as a docker image.** In fact, container runtime itself is one of the test targets of the node conformance test. We should not rely on it to run the test. However, container image is actually the easiest way to package and ship your program to different platforms. To compromise, we decided to:
  - Ship a docker image version for most users.
  - Build the test suite into a single binary to make it easy for users who don't like the container runtime dependency to build and run the test themselves.

@dchen1107 @vishh 
/cc @kubernetes/sig-node

**Comment by vishh** [comment]
I'd prefer including kubelet also in the static binary and unifying conformance testing and node e2e. We can run kubelet as part of testing in node e2e and not run in the case of conformance testing. The default can be that of not running kubelet.

**Comment by Random-Liu** [comment]
> I'd prefer including kubelet also in the static binary and unifying conformance testing and node e2e. We can run kubelet as part of testing in node e2e and not run in the case of conformance testing. The default can be that of not running kubelet.

There are several questions about this:
**1) Should we statically linked in kubelet?**
I prefer not.
* For node conformance test, we assume user should have set up kubelet on the node, so no need to carry one ourselves.
* For node e2e test, I don't think test statically linked kubelet is testing "real kubelet". We should test kubelet binary generated by our build process, started by our `cmd/kubelet`, deploy it and test it, which I think is "real kubelet". We can statically link etcd, apiserver and namespace controller, because our purpose is not testing them, they are just part of test framework. However, kubelet is different.

If the theory makes sense, then both node conformance test and node e2e test don't need to link in kubelet. :)

**2) Should we start kubelet in the test suite?**
One problem of starting kubelet in the test suite is that we plan to run node conformance test in container. However, currently kubelet has to run on the host. There is no good way to start a native process from inside a container (maybe ssh or something). So we can't start kubelet in test suite for node conformance test.
For node e2e test, we can start kubelet in test suite, if we don't containerize the test. However, in this way node conformance and e2e test are not unified, because we've to start kubelet in one and not in another.

If we start kubelet outside of the test suite, the framework could be shared in node conformance test and node e2e test.

For node conformance test:
Step 1: User set up the node, including kubelet, cni etc. (In the future, using installer)
------ Shared from here ------
Step 2: Run node test script.
* If kubelet is not started, the script starts kubelet with default flags.
* If kubelet is running but is pointed to local apiserver, the script restarts kubelet keeping all flags except apiserver address.
Step 2: The node test script starts node test container with current kubelet flag as one of the arguments.

For node e2e test:
Step 1: Node e2e framework ssh and set up the node, including kubelet, cni etc. (In the future, using installer)
Step 2: Run kubelet with flags we want to test with.
------ Shared from here ------
Step 3: Run the same node test script.
Step 4: The node test script starts node test container with current kubelet flag as one of the arguments.

In this way, we can:
1) Make sure kubelet is installed in the same way with user in the future.
2) Unify node conformance test and node e2e test.

@vishh @dchen1107 WDYT?

**Comment by dims** [comment]
This needs to be triaged as a release-blocker or not for 1.5 @Random-Liu

**Comment by dchen1107** [comment]
All work items listed in this issue were addressed. There are several enhancement related to the node conformance test being tracked through #30122.
